pyvera/__init__.py

```diff
@@ -878,10 +893,89 @@ def is_locked(self, refresh=False):
         Refresh data from Vera if refresh is True, otherwise use local cache.
         Refresh is only needed if you're not using subscriptions.
         """
+        try:
+            if refresh:
+                self.refresh()
+            return self.get_value("locked") == '1'
+        except (TypeError, ValueError):
```

Comment:
Do you know the circumstances where the fallback is needed? Is it just on startup or some sort of race condition?

Reply:
If I remember correctly, there was no race condition or circumstance that I encountered where the fallback was used.  I believe I added it there for the sake of completeness since it appeared to actually be the only documented “micasa” way of getting the locked status

Reply:
Want to throw my 2p in here. I think using get_complex_value("Status") is superior to get_value("locked") here and the former should be preferred, not the fallback. The reason is that in the absence of Refresh=True, they are about equal, but with Refresh=True, the sdata refresh needed with get_value() requires fetching of data for ALL devices (an sdata request always returns full data) and then a linear search through those results to find the subject device, where the complex value refresh uses a more efficient single-device ("status" with DeviceNum parameter) request.

However, and I'm willing to tackle this as a separate PR, it's also possible that the sdata refresh() could be made much more efficient by incorporating the use of get_changed_devices() and applying the results to the currently cached data. Additionally, if the polling/subscription loop is running, it could be updating the pyvera's cached sdata as it works, something it does not currently seem to do. Right now, the polling/subscription loop could easily provide a different value to a client callback than a get_value() call (without prior refresh) would return for the same name, and that should not be the case. Addressing this would make the need for calls with Refresh=True much less necessary/likely, and reduce load on the Vera and efficiency for the client.

Reply:
I hadn’t considered that, we might see some good improvements with that scheme.  We may make a sacrifice of commonality between devices for better performance right?
If you write up the pr, I’d be willing to help out where I can.  You could start with some pseudo code or an example to help better explain your vision and then we could latch on

Reply:
Given the fallback isn't needed - I suggest we remove it to match the rest of the code base for now. If we want to move away from the `sdata` calls we can do this as another PR.

For now I think `sdata` is the better approach because it works with the subscription mechanism, and most users of pyvera are using it as part of HA - and so will benefit from realtime updates. I accept that there may be a better way (PR welcome).

This assumes that the subscription mechanism works correctly for locks (you might want to check that the ugly code here (https://github.com/pavoni/pyvera/blob/master/pyvera/subscribe.py#L78) works and is still needed, and subscriptions work properly for locks.